common/channel/src/libra_channel.rs

```diff
@@ -0,0 +1,109 @@
+//! libra_channel provides an mpsc channel which has two ends `libra_channel::Receiver`
+//! and `libra_channel::Sender` similar to existing mpsc data structures.
+//! What makes it different from existing mpsc channels is that we have full control
+//! over how the internal queueing in the channel happens and how we schedule messages
+//! to be sent out from this channel.
+//! Note that libra_channel does not provide an implementation of this internal queueing
+//! mechanism, but provides a trait (`MessageQueue`) which the consumers of this channel
+//! can use to create libra_channels' according to their needs.
+
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::Waker;
+
+use futures::async_await::FusedStream;
+use futures::stream::Stream;
+use futures::task::Context;
+use futures::Poll;
+
+/// MessageQueue is a trait which provides a very simple set of methods for implementing
+/// a queue which will be used as the internal queue libra_channel.
+pub trait MessageQueue {
+    /// The actual type of the messages stored in this MessageQueue
+    type Message;
+
+    /// Push a message to this queue
+    fn push(&mut self, message: Self::Message);
+
+    /// Pop a message from this queue
+    fn pop(&mut self) -> Option<Self::Message>;
+}
+
+/// SharedState is a data structure private to this module which is
+/// shared by the sender and receiver.
+struct SharedState<T> {
+    /// The internal queue of messages in this Channel
+    internal_queue: T,
+
+    /// Waker is needed so that the Sender can notify the task executor/scheduler
+    /// that something has been pushed to the internal_queue and it ready for
+    /// consumption by the Receiver and then the executor/scheduler will wake up
+    /// the Receiver task to process the next item.
+    waker: Option<Waker>,
+}
+
+/// The sending end of the libra_channel.
+#[derive(Clone)]
+pub struct Sender<T> {
+    shared_state: Arc<Mutex<SharedState<T>>>,
+}
+
+impl<T: MessageQueue> Sender<T> {
+    /// This adds the message into the internal queue data structure. This is a non-blocking
+    /// synchronous call.
+    /// TODO: We can have this return a boolean if the queue of a validator is capacity
+    pub fn put(&mut self, message: <T as MessageQueue>::Message) {
+        let mut shared_state = self.shared_state.lock().unwrap();
+        shared_state.internal_queue.push(message);
+        if let Some(w) = shared_state.waker.take() {
+            w.wake();
+        }
+    }
```

Comment:
You probably want to use the futures::Sink trait instead of a custom put method. One thing that using that trait would prompt is thinking about what happens if you can't enqueue a message into the channel like in the case where the receiver has been dropped.

Reply:
We wanted to implement a synchronous api on the sender side of this channel, hence I chose not to implement the Sink trait. But I'll add another PR to keep track of whether the Receiver has been dropped.

Reply:
I don't have the full context so my comments may be out of place, but i would still recommend some caution with mixing sync and async api's. What happens if the channel is full? do you block? do you throw the message away? do you wait till there is room? if you use an async api then you can do any of the options i've listed but if you use a sync api then you cannot ever block and wait till there is room in the queue (assuming this channel will be used inside of a Future) because blocking in a future like that could cause deadlocks.

Reply:
When the channel is full, we throw drop messages instead of waiting for the channel to have room. We wanted to do this to limit the number of messages each validator can queue up.

Reply:
> i would still recommend some caution with mixing sync and async api's

Mixing sync and async API actually seem to be valid pattern.
For example sending to async oneshot channel is a non-async operation. So basically if operation does not need to block it makes sense to make it non-async